eyelid-match-ops/src/primitives/yashe.rs

```diff
@@ -0,0 +1,119 @@
+//! Implementation of YASHE cryptosystem
+//! `<https://eprint.iacr.org/2013/075.pdf>`
+
+use crate::primitives::poly::{inverse, modular_poly::Poly, Coeff};
+use ark_ff::{One, UniformRand, Zero};
+use rand::rngs::ThreadRng;
+use rand_distr::{Distribution, Normal};
+
+pub mod test;
+
+/// Yashe parameters
+#[derive(Clone)]
+pub struct YasheParams {
+    /// Plaintext coefficient modulus
+    pub t: u64,
+    /// Standard deviation
+    pub delta: f64,
+}
+
+/// Yashe scheme
+pub struct Yashe<const MAX_POLY_DEGREE: usize> {
+    /// Cryptosystem parameters
+    params: YasheParams,
+}
+
+/// Private key struct
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct PrivateKey<const MAX_POLY_DEGREE: usize> {
+    /// Sampled with small coefficients (and invertible)
+    f: Poly<MAX_POLY_DEGREE>,
+    /// The inverse of f
+    finv: Poly<MAX_POLY_DEGREE>,
+    /// Private key
+    priv_key: Poly<MAX_POLY_DEGREE>,
+}
+
+/// Public key struct
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct PublicKey<const MAX_POLY_DEGREE: usize> {
+    /// Public key
+    h: Poly<MAX_POLY_DEGREE>,
+}
+
+impl<const MAX_POLY_DEGREE: usize> Yashe<MAX_POLY_DEGREE> {
+    /// Yashe constructor
+    pub fn new(params: YasheParams) -> Self {
+        Self { params }
+    }
+
+    /// Generate the private key
+    pub fn generate_private_key(&self, rng: ThreadRng) -> PrivateKey<MAX_POLY_DEGREE> {
+        loop {
```

Comment:
If the RNG is broken or there's a bug in the code, key generation could become an infinite loop. (Or extremely slow.)

Should we loop a limited number of times instead, and return an error if that limit is exceeded?

Reply:
It is a high probability of getting good key material right in the first pass.

Reply:
> It is a high probability of getting good key material right in the first pass.

What is the approximate probability? (order of magnitude is close enough)
That impacts how we deal with failures.

Did you want to deal with this in another ticket?
If so, we can just add another TODO in this PR, like this (you might need to load more comments to see it):
https://github.com/Inversed-Tech/eyelid/pull/69#discussion_r1583882681

Reply:
This is now #74